File: leet/easy/reverselinkedlist.py
# https://leetcode.com/problems/reverse-linked-list/
from typing import Optional

from leet.structures import ListNode, list_to_linked_list, linked_list_to_list


class Solution:
    @linked_list_to_list
    @list_to_linked_list
    def reverse_list(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next:
            return head

        rev_head = None
        while head:
            next = head.next
            head.next = rev_head
            rev_head = head
            head = next
        return rev_head

    @linked_list_to_list
    @list_to_linked_list
    def reverse_list_dirty(self, head: Optional[ListNode]) -> (Optional[ListNode], Optional[ListNode]):
        if not head or not head.next:
            return head
        prev = None
        tail = head
        next = head.next
        while head:
            head.next = prev
            prev = head
            head = next

            if next:
                next = next.next
        return prev, tail

File: src/leet/easy/reverselinkedlist.rs
// https://leetcode.com/problems/reverse-linked-list/

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode {
            next: None,
            val
        }
    }
}

pub struct Solution;

impl Solution {
    pub fn reverse_list(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        if let Some(start) = head {
            let mut head = None;
            let mut current = start;
            loop {
                current.next = head;
                head = Some(current);
            }
            return head;  

        }
        return None;
    }
}